### PR TITLE
treewide get rid of go 1.12

### PR DESCRIPTION
**Pull request by c00w**
###### Motivation for this change
Getting rid of go1.12 dependencies so we can drop go1.12. Also this accidentally got dropped out of #81159

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by c00w** [comment]
@worldofpeace Sorry - these got dropped out of #81159

**Comment by c00w** [comment]
And I got excited, so this PR now completely removes go 1.12 (and almost go1.13)

File: pkgs/applications/networking/cluster/openshift/default.nix
Comment (vdemeester) [inline pkgs/applications/networking/cluster/openshift/default.nix]: :heart_eyes:

**Review by vdemeester — APPROVED** [review]
Nice :clap: 
LGTM :upside_down_face:

File: pkgs/servers/consul/default.nix
Comment (nh2) [inline pkgs/servers/consul/default.nix]: Could you comment (here, or even better in the commit message) whether this restriction is now lifted?

File: pkgs/servers/consul/default.nix
Comment (nh2) [inline pkgs/servers/consul/default.nix]: Could you comment (here, or even better in the commit message) whether/why consul no longer needs this passed to know its own git version?

File: pkgs/servers/consul/default.nix
Comment (nh2) [inline pkgs/servers/consul/default.nix]: Could this have a comment by what means one discovers what has to go in here?

**Review by nh2 — COMMENTED** [review]
Thanks! I've put in some small requests/comments on Consul.

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: No idea, but in general this is universally true for all software. If you try and build at a broken commit it won't work - if consul fails to check in js assets and it fails, then the commit was bad. I don't see why we need to call this out for advanced users who are modifying nix expressions to change which revision something builds at.

Also the linked PR doesn't have anything to do w/ consul - that's vault (which is by the same people), but is definitely a different piece of software.

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: Interesting question - consul does something interesting (or broken) here, so if you let the default try (which is build everything) it errors due to spanning multiple modules. These are the two paths we need to build to match the binaries that were in the existing modules. But in general, play with it until it builds is the heuristic I did. (followed by double check the same binaries are in the output).

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: It builds without it - I just deleted all customizations and produced the minimum building version.

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: I don't think this is a big deal though, added the comments back.

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: Added a comment which may help future people.

File: pkgs/servers/consul/default.nix
Comment (c00w) [reply pkgs/servers/consul/default.nix]: Added lines back in because I don't have strong opinions.

File: pkgs/servers/consul/default.nix
Comment (nh2) [reply pkgs/servers/consul/default.nix]: > No idea, but in general this is universally true for all software.

That is not my experience -- for most projects the release tarballs are just snapshots of the actual repository, and it is very common for nix users to modify `src` in `overrideAttrs` with a newer version from git; consul is behaving particularly badly here, which is why I found this comment helpful.

> Also the linked PR doesn't have anything to do w/ consul - that's vault

The PR diff is about vault, but most nixpkgs maintainers discussion about the consul UI has been in that thread (search for `consul` on the page), that's why got linked.

> I don't think this is a big deal though, added the comments back.

Thanks!

File: pkgs/servers/consul/default.nix
Comment (nh2) [reply pkgs/servers/consul/default.nix]: > It builds without it - I just deleted all customizations and produced the minimum building version.

I don't think that is a good approach in general -- if you check out the `git blame` it shows that this fixed real issues (#24606; users expect that a program's `--version` works).

I do think that commit should have had a comment though -- my bad!

> Added lines back in because I don't have strong opinions.

:+1:

**Review by nh2 — APPROVED** [review]
LGTM for the `consul` change.

**Comment by c00w** [comment]
@kalbasit @worldofpeace  Any chance you could review + merge?

File: pkgs/applications/networking/cluster/prow/default.nix
Comment (kalbasit) [inline pkgs/applications/networking/cluster/prow/default.nix:6]: This was already done in #84055. You should probably rebase.

**Review by kalbasit — CHANGES_REQUESTED** [review]

**Comment by worldofpeace** [comment]
@c00w Yeah a rebase is needed. (and resolve conflicts). Merge seems good to me.

File: pkgs/applications/networking/cluster/prow/default.nix
Comment (c00w) [reply pkgs/applications/networking/cluster/prow/default.nix]: Should be sorted now.

**Comment by c00w** [comment]
Rebased - everything still passes with 
```
nix-build -A lnd -A perkeep -A openshift -A packr -A consul -A scaleway-cli -A prometheus-varnish-exporter -A curlie -A grpcui -A ipfs -A nomad -A gobetween -A cockroachdb -A gotestsum -A magnetico -A tendermint -A cri-o default.nix
```

**Comment by c00w** [comment]
@kalbasit @worldofpeace  Should be good for another round of review :)

**Comment by worldofpeace** [comment]
@GrahamcOfBorg eval

**Review by kalbasit — APPROVED** [review]
LGTM. May need another rebase to fix the eval but it looks good overall. Before you merge, we should ask the bot to build all affected packages on Linux/Mac.

**Comment by c00w** [comment]
Thanks :) AFAIK I don't have permission to merge, or talk to the bot, so someone needs to fix it.

The eval failure is actually a interesting bug in grahamcofborg, where if you delete a file, it fails. And since by design I'm deleting to go1.12 files, it gets hit.

**Comment by worldofpeace** [comment]
The eval error is
```
while evaluating the attribute 'deps' of the derivation 'bazel-watcher-0.10.3' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/pkgs/build-support/build-bazel-package/default.nix:42:10:
while evaluating the attribute 'nativeBuildInputs' of the derivation 'bazel-watcher-0.10.3-deps' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/pkgs/build-support/build-bazel-package/default.nix:45:5:
while evaluating 'getOutput' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/lib/attrsets.nix:464:23, called from undefined position:
while evaluating anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/pkgs/stdenv/generic/make-derivation.nix:133:17, called from undefined position:
while evaluating 'callPackageWith' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/lib/customisation.nix:117:35, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/pkgs/top-level/all-packages.nix:8562:13:
while evaluating 'makeOverridable' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/lib/customisation.nix:67:24, called from /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/lib/customisation.nix:121:8:
getting status of '/var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/hetzner-sb65-1083082-hel1-dc2/pkgs/development/compilers/go/1.12.nix': No such file or directory
```

Using `rg` I find
```
pkgs/top-level/all-packages.nix
8515:  go_1_12 = callPackage ../development/compilers/go/1.12.nix ({
```

we need to actually remove `go_1_12` in `all-packages`nix` and add an alias in `alias.nix` to `throw`.

**Comment by c00w** [comment]
Good catch - deleted missing reference in go_1_12. I don't see go_1_11 in aliases.nix, so it seems like it's unlikely to be needed?

**Comment by c00w** [comment]
Rebase on top of the PR removing go1.13

**Comment by worldofpeace** [comment]
> I don't see go_1_11 in aliases.nix, so it seems like it's unlikely to be needed? 

There isn't really a likelihood of something needing an alias or a `throw`, it's pretty definite if something is renamed or either removed. For the end-user, having something like

```
buildGo112Package = throw "buildGo112Package has been removed from nixpkgs.";
```

is slightly more helpful than an attribute just not existing.

**Comment by worldofpeace** [comment]
Seeing this hasn't been done regularly we could ignore that for this particular PR, but I know these help people.

**Comment by c00w** [comment]
@worldofpeace Added.

**Comment by c00w** [comment]
Updated PR with latest master due to a conflict.

File: pkgs/top-level/aliases.nix
Comment (worldofpeace) [inline pkgs/top-level/aliases.nix:63]: shouldn't we also do `go_1_12`?

Comment (c00w) [reply pkgs/top-level/aliases.nix:63]: *shrug* sure. I doubt anyone will ever use it, but done.

**Comment by c00w** [comment]
Rebased on master.

**Comment by worldofpeace** [comment]
Thank you for your patience @c00w and your contribution :sparkles:

**Pull request merged**
